server/db/mongodb/adapter.go

```diff
@@ -0,0 +1,2163 @@
+	if err != nil {
+		return nil, err
+	}
+	defer cur.Close(a.ctx)
+
+	var uids []t.Uid
+	var userId map[string]string
+	for cur.Next(a.ctx) {
+		if err := cur.Decode(&userId); err != nil {
+			return nil, err
+		}
+		uids = append(uids, t.ParseUid(userId["_id"]))
+	}
+	return uids, nil
+}
+
+// UserUpdate updates user record
+func (a *adapter) UserUpdate(uid t.Uid, update map[string]interface{}) error {
+	// to get round the hardcoded "UpdatedAt" key in store.Users.Update()
+	update = normalizeUpdateMap(update)
+
+	_, err := a.db.Collection("users").UpdateOne(a.ctx, b.M{"_id": uid.String()}, b.M{"$set": update})
+	return err
+}
+
+// UserUpdateTags adds, removes, or resets user's tags
+func (a *adapter) UserUpdateTags(uid t.Uid, add, remove, reset []string) ([]string, error) {
+	// Compare to nil vs checking for zero length: zero length reset is valid.
+	if reset != nil {
+		// Replace Tags with the new value
+		return reset, a.UserUpdate(uid, map[string]interface{}{"tags": reset})
+	}
+
+	var user t.User
+	err := a.db.Collection("users").FindOne(a.ctx, b.M{"_id": uid.String()}).Decode(&user)
+	if err != nil {
+		return nil, err
+	}
+
+	// Mutate the tag list.
+	newTags := user.Tags
+	if len(add) > 0 {
+		newTags = union(newTags, add)
+	}
+	if len(remove) > 0 {
+		newTags = diff(newTags, remove)
+	}
+
+	update := map[string]interface{}{"tags": newTags}
+	if err := a.UserUpdate(uid, update); err != nil {
+		return nil, err
+	}
+
+	// Get the new tags
+	var tags map[string][]string
+	findOpts := mdbopts.FindOne().SetProjection(b.M{"tags": 1, "_id": 0})
+	err = a.db.Collection("users").FindOne(a.ctx, b.M{"_id": uid.String()}, findOpts).Decode(&tags)
+	if err != nil {
+		return nil, err
+	}
+
+	return tags["tags"], nil
+}
+
+// UserGetByCred returns user ID for the given validated credential.
+func (a *adapter) UserGetByCred(method, value string) (t.Uid, error) {
+	var userId map[string]string
+	err := a.db.Collection("credentials").FindOne(a.ctx,
+		b.M{"_id": method + ":" + value},
+		mdbopts.FindOne().SetProjection(b.M{"user": 1, "_id": 0}),
+	).Decode(&userId)
+	if err != nil {
+		if err == mdb.ErrNoDocuments {
+			return t.ZeroUid, nil
+		}
+		return t.ZeroUid, err
+	}
+
+	return t.ParseUid(userId["user"]), nil
+}
+
+// UserUnreadCount returns the total number of unread messages in all topics with
+// the R permission.
+func (a *adapter) UserUnreadCount(uid t.Uid) (int, error) {
+	pipeline := b.A{
+		b.M{"$match": b.M{"user": uid.String()}},
+		// Join documents from two collection
+		b.M{"$lookup": b.M{
+			"from":         "topics",
+			"localField":   "topic",
+			"foreignField": "_id",
+			"as":           "fromTopics"},
+		},
+		// Merge two documents into one
+		b.M{"$replaceRoot": b.M{"newRoot": b.M{"$mergeObjects": b.A{b.M{"$arrayElemAt": b.A{"$fromTopics", 0}}, "$$ROOT"}}}},
+
+		b.M{"$match": b.M{
+			"deletedat": b.M{"$exists": false},
+			// Filter by access mode
+			"modewant":  b.M{"$bitsAllSet": b.A{1}},
```

Comment:
b.A{1} - it reads you want bit 1 to be set.
Is the read permission not bit 0? Or am I missing something here?

Reply:
This `match` returns all documents where `modewant`s **bit-1** set to 1 (have read permission)

Reply:
I see. Thanks for explaining.